### DISCUSSION TITLE
Transform from WGS84 to VN2000

### DISCUSSION THREAD

**Discussion by thothai94** [ngageoint/projections-ios#4] [Q&A] [answered]
Hello!
As title of this discussion. I would like to define a PROJProjection by proj4 params and make a transformation from WGS84 to VN2000 in my Swift project
Please take a look at my Swift code below:
```
let location: CLLocationCoordinate2D = CLLocationCoordinate2D(latitude: 10.785287181209878, longitude: 106.69332489927747)

let wgs84: PROJProjection = PROJProjectionFactory.projection(withName: "EPSG:4326")
let vn2000:PROJProjection = PROJProjectionFactory.projection(withAuthority: PROJ_AUTHORITY_EPSG, andIntCode: 9210, andParams: "+proj=tmerc +lat_0=0 +lon_0=105.75 +k=0.9999 +x_0=500000 +y_0=0 +ellps=WGS84 +towgs84=-191.90441429,-39.30318279,-111.45032835,-0.00928836,0.01975479,-0.00427372,0.252906278 +units=m +no_defs")

let transform: PROJProjectionTransform = vn2000.transformation(with: wgs84)
let transformed: CLLocationCoordinate2D = transform.transform(location)
let utmTransformed = transformed.utmCoordinate() //use UTMConvertion to convert transformed

print("transformed: \(transformed) - \(utmTransformed)")
```
But the result seem to be wrong with the online transform tool at: https://epsg.io/transform#s_srs=4326&t_srs=9210&x=106.6933322&y=10.7852875

Please tell me how to define a PROJProjection from proj4 params and help me solve this issue

**Comment by bosborn** [comment]
Swap the [transform creation](https://ngageoint.github.io/projections-ios/docs/api/Classes/PROJProjection.html#/c:objc(cs)PROJProjection(im)transformationWithProjection:) direction so it is from 'wgs84' to 'vn2000' (vs 'vn2000'->'wgs84').

`let transform: PROJProjectionTransform = wgs84.transformation(with: vn2000)`

**Reply by thothai94** [reply]
Thank you for your answer.
I tried again with your suggestion but the result did not match with the online tool
I think my creation of VN2000 PROJProjection can be wrong. Can you tell me the right method and input infomations to create an PROJProjection by proj4 params string like this: 
"+proj=tmerc +lat_0=0 +lon_0=105.75 +k=0.9999 +x_0=500000 +y_0=0 +ellps=WGS84 +towgs84=-191.90441429,-39.30318279,-111.45032835,-0.00928836,0.01975479,-0.00427372,0.252906278 +units=m +no_defs"

**Reply by bosborn** [reply, marked as answer]
```
let location: CLLocationCoordinate2D = CLLocationCoordinate2D(latitude: 10.785287181209878, longitude: 106.69332489927747)

let wgs84: PROJProjection = PROJProjectionFactory.projection(withName: "EPSG:4326")
let vn2000:PROJProjection = PROJProjectionFactory.projection(withAuthority: PROJ_AUTHORITY_EPSG, andIntCode: 9210, andParams: "+proj=tmerc +lat_0=0 +lon_0=105.75 +k=0.9999 +x_0=500000 +y_0=0 +ellps=WGS84 +towgs84=-191.90441429,-39.30318279,-111.45032835,-0.00928836,0.01975479,-0.00427372,0.252906278 +units=m +no_defs")

let transform: PROJProjectionTransform = wgs84.transformation(with: vn2000)
let transformed: CLLocationCoordinate2D = transform.transform(location)

print("transformed: \(transformed)")
```
`transformed: CLLocationCoordinate2D(latitude: 1192867.7008336019, longitude: 602966.4073691196)`

Appears to be reasonably close with [epsg.io's tool](https://epsg.io/transform#s_srs=4326&t_srs=9210&x=106.6933322&y=10.7852875)
![image](https://user-images.githubusercontent.com/11667761/218461538-bd5ede7d-dd5e-4e27-b00c-551c80249eab.png)

Matches exactly with higher precision using [MyGeodata's tool](https://mygeodata.cloud/cs2cs/)
![image](https://user-images.githubusercontent.com/11667761/218461944-31e1ffca-bf65-4b3e-ba21-0d1b86770b1a.png)
